### ISSUE TITLE
When Async.complete() is called and not all bytes are written no error is issued and the channel is not closed

### ISSUE THREAD

**Issue by laeubi** [eclipse/jetty.project#1034]
This is related to this news-list post: https://dev.eclipse.org/mhonarc/lists/jetty-users/msg07332.html

If you are setting a Content-Length and then use WriteListener to write bytes no error is issued and the channel is not closed.

You can find an example here: https://github.com/laeubi/jetty_test_issues/blob/master/jetty.test.issues/src/jetty/async/test/IncompleteWriteListenerServlet.java
The example sets the content length to N and then writed N - 10 bytes then completes the AsyncContext.

What happens:

```
 Request: /iwl
 Setting Content-Length: 10000
 DebugWriteListener.onWritePossible()
 Write 9990 bytes...
 do complete...
 DebugAsyncListener.onComplete()
```

Using a browser (Firefox in my case) the Request just hangs on client side, after 30 seconds (seems not to relate to the AsyncTimeout, see example) the connection terminates.

What is expected:
On Serverside onError is called, the client connection is terminated immediately.

I tried to check case b) that Simone Bordet has suggested with https://github.com/laeubi/jetty_test_issues/blob/master/jetty.test.issues/src/jetty/async/test/TimeoutWriteListenerServlet.java this results in an 

```
Request: /twl
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugWriteListener.onWritePossible()
Stream is not ready...
DebugAsyncListener.onTimeout()
DebugWriteListener.onError(org.eclipse.jetty.io.EofException: Async close)
DebugAsyncListener.onComplete()
Request: /btwl
DebugWriteListener.onWritePossible()
```

and the client connection terminates immediately as expected.

**Comment by joakime** [comment]
Seems like this related to Issue #264

**Comment by gregw** [comment]
I don't think #264 is related.  

This has taken a long time to look at, but I have reproduced and it is sub-optimal behaviour... even if a programming error.

**Comment by gregw** [comment]
reproduced in 9.3, but fixed in 9.4 by #1313

**Comment by laeubi** [comment]
Great news, thanks for fixing!
